psi4/driver/procrouting/scf_proc/subclass_methods.py

```diff
@@ -119,3 +122,106 @@ def _ROHF_diis(self, Dnorm):
 core.RHF.diis = _RHF_diis
 core.UHF.diis = core.CUHF.diis = _UHF_diis
 core.ROHF.diis = _ROHF_diis
+
+def _UHF_stability_analysis(self):
+    # => Validate options <=
+    # TODO: Stability analysis is supported for any functional UKS functional where its one-
+    # and two-body Hamiltonian matrix-vector products are implemented. This is true for LDA
+    # at least, but probably not other functionals. This restriction exists for now because
+    # we've always had it, but we should lift it as much as we can and implement more matrix-
+    # vector products so we can lift it further.
+    # TODO: It should be up to the SolverEngine to validate whether it can do Hx products for the input wfn.
+    if self.functional().is_gga() or self.functional().is_meta():
+        raise ValidationError("Stability analysis not yet supported for non-LDA functionals.")
+
+    # => Prep options for eigenvector solver <=
+    if not self.options().has_changed("SOLVER_ROOTS_PER_IRREP"):
+        roots = [self.options().get_int("SOLVER_N_ROOT")] * self.nirrep()
+    else:
+        roots = self.options().get_int_vector("SOLVER_ROOTS_PER_IRREP")
+        if len(roots) != wfn.nirrep():
+            raise ValidationError(f"SOLVER_ROOTS_PER_IRREP specified {wfn.nirrep()} irreps, but there are {len(roots)} irreps.")
+    r_convergence = self.options().get_double("SOLVER_CONVERGENCE")
+    if not self.options().has_changed("SOLVER_N_GUESS"):
+        # Below formula borrowed from TDSCF code.
+        max_vecs_per_root = int(-np.log10(r_convergence) * 50)
+    else:
+        max_vecs_per_root = self.options().get_int("SOLVER_N_GUESS")
+    engine = TDUSCFEngine(self, ptype="hess")
+
+    # => Compute eigenvectors and do trivial data processing <=
+    eval_sym = core.Matrix("SCF STABILITY EIGENVALUES", core.Dimension(roots), core.Dimension([1] * self.nirrep()))
+    current_eigenvalue = None
+    unstable = False
+    for h, nroot in enumerate(roots):
+        if not nroot: continue
+        # The below line changes the guess the engine generates, which controls the final states.
+        # This selects for eigenvectors of irrep h.
+        engine.reset_for_state_symm(engine.G_gs ^ h)
+        ret = solvers.davidson_solver(engine=engine,
+                                      nroot=nroot,
+                                      guess=engine.generate_guess(nroot * 4),
```

Comment:
I have trouble understanding the hard coded `* 4` but I see it is taken from the TDSCF code. Isn't there a risk of generating a too many guess vectors for a small system?

Reply:
No. The engine's `generate_guess` function will generate the _minimum_ of the requested number of guesses and the possible number of guesses.